File: lang-core/src/test/java/com/jaitlapps/differ/lexical/impl/DifferLexicalAnalyzerTest.java
package com.jaitlapps.differ.lexical.impl;

import com.jaitlapps.differ.lexical.LexicalAnalyzer;
import com.jaitlapps.differ.model.ExpressionType;
import com.jaitlapps.differ.model.KeywordType;
import com.jaitlapps.differ.model.Token;
import com.jaitlapps.differ.model.TokenType;
import com.jaitlapps.differ.reader.impl.StringReader;
import org.junit.Test;

import static org.junit.Assert.*;

public class DifferLexicalAnalyzerTest {
    @Test
    public void testKeywordBegin() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("Начало"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Keyword);
        assertEquals(token.getKeywordType(), KeywordType.Begin);
    }

    @Test
    public void testTokenTypeUnknown() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("44rrr"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Unknown);
    }

    @Test
    public void testKeywordReal() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("Третье"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Keyword);
        assertEquals(token.getKeywordType(), KeywordType.Third);
    }

    @Test
    public void testKeywordInteger() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("Первое"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Keyword);
        assertEquals(token.getKeywordType(), KeywordType.First);
    }

    @Test
    public void testKeywordBeginEof() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("  Начало     "));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Keyword);
        assertEquals(token.getKeywordType(), KeywordType.Begin);

        token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Eof);

        token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Eof);
    }

    @Test
    public void testInteger125() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("125"));
        final int actual = 125;

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Integer);
        assertEquals(token.getValue(), actual);
    }

    @Test
    public void testVariableAs456r() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("as456r"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Variable);
        assertEquals(token.getCapitalizedString(), "as456r");
    }

    @Test
    public void testVariable4ry44ed() {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("4ry44ed"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Unknown);
    }

    @Test
    public void test_variable_ry4ed()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("ry4#ed"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Unknown);
    }

    @Test
    public void test_label_1234()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("1234:"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Label);
    }

    @Test
    public void test_label_ry4ed()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("ry4ed:"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Unknown);
    }

    @Test
    public void test_label_23458()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("23458"));

        Token token = lexicalAnalyzer.nextToken();

        assertNotEquals(token.getTokenType(), TokenType.Label);
    }

    @Test
    public void test_label_negative_23458()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("-23458"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getExpressionType(), ExpressionType.Subtraction);

        token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Integer);
    }

    @Test
    public void test_expression_add()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("+"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Expression);
        assertEquals(token.getExpressionType(), ExpressionType.Addition);
    }

    @Test
    public void test_expression_sub()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("-"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Expression);
        assertEquals(token.getExpressionType(), ExpressionType.Subtraction);
    }

    @Test
    public void test_expression_multi()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("*"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Expression);
        assertEquals(token.getExpressionType(), ExpressionType.Multiplication);
    }

    @Test
    public void test_expression_div()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("/"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Expression);
        assertEquals(token.getExpressionType(), ExpressionType.Division);
    }

    @Test
    public void test_expression_inv()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("^"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Expression);
        assertEquals(token.getExpressionType(), ExpressionType.Involution);
    }

    @Test
    public void test_equal()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("="));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Equal);
    }

    @Test
    public void test_open_bracket()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("["));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getExpressionType(), ExpressionType.OpenBracket);
    }

    @Test
    public void test_close_bracket()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("]"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getExpressionType(), ExpressionType.CloseBracket);
    }

    @Test
    public void test_real_1()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("43.332"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Real);
    }

    @Test
    public void test_real_2()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader("43."));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Unknown);
    }

    @Test
    public void test_real_3()
    {
        LexicalAnalyzer lexicalAnalyzer = new DifferLexicalAnalyzer(new StringReader(".4534"));

        Token token = lexicalAnalyzer.nextToken();

        assertEquals(token.getTokenType(), TokenType.Unknown);
    }
}

File: lang-core/src/test/kotlin/com/jaitlapps/differ/lexical/impl/DifferLexicalAnalyzerTest.kt
package com.jaitlapps.differ.lexical.impl

import com.jaitlapps.differ.model.KeywordType
import com.jaitlapps.differ.model.MethodType
import com.jaitlapps.differ.model.SymbolType
import com.jaitlapps.differ.model.TokenType
import com.jaitlapps.differ.model.token.KeywordToken
import com.jaitlapps.differ.model.token.MethodToken
import com.jaitlapps.differ.model.token.NumberToken
import com.jaitlapps.differ.model.token.SymbolToken
import com.jaitlapps.differ.reader.impl.StringReader
import org.junit.Test
import kotlin.test.assertEquals

class DifferLexicalAnalyzerTest {
    @Test
    fun testTokenTypeUnknown(){
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("йапрограмко"))
        val token = lexicalAnalyzer.nextToken()

        assertEquals(TokenType.Unknown, token.tokenType)
    }

    @Test
    fun testKeywordProgram() {
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("Программа"))

        val token:KeywordToken = lexicalAnalyzer.nextToken() as KeywordToken

        assertEquals(TokenType.Keyword, token.tokenType)
        assertEquals(KeywordType.Program, token.keywordType)

        val nextToken = lexicalAnalyzer.nextToken()
        assertEquals(TokenType.Eof, nextToken.tokenType)
    }

    @Test
    fun testKeyWordMethod() {
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("Метод Эйлера;"))

        val token:KeywordToken = lexicalAnalyzer.nextToken() as KeywordToken

        assertEquals(TokenType.Keyword, token.tokenType)
        assertEquals(KeywordType.Method, token.keywordType)

        val methodToken = lexicalAnalyzer.nextToken() as MethodToken
        assertEquals(TokenType.Method, methodToken.tokenType)
        assertEquals(MethodType.Eiler, methodToken.methodType)

        val symbolToken = lexicalAnalyzer.nextToken() as SymbolToken
        assertEquals(TokenType.Symbol, symbolToken.tokenType)
        assertEquals(SymbolType.Semicolon, symbolToken.symbolType)
    }

    @Test
    fun testKeyWordDefinition() {
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("Коэффициенты a = 4;"))

        val token:KeywordToken = lexicalAnalyzer.nextToken() as KeywordToken

        assertEquals(TokenType.Keyword, token.tokenType)
        assertEquals(KeywordType.Coefficient, token.keywordType)

        val coefficientToken = lexicalAnalyzer.nextToken()

        assertEquals(TokenType.Coefficient, coefficientToken.tokenType)
        assertEquals("a", coefficientToken.word.word)

        val equalToken = lexicalAnalyzer.nextToken() as SymbolToken

        assertEquals(TokenType.Symbol, equalToken.tokenType)
        assertEquals(SymbolType.Equal, equalToken.symbolType)

        val numberToken = lexicalAnalyzer.nextToken() as NumberToken

        assertEquals(TokenType.Integer, numberToken.tokenType)
        assertEquals(4, numberToken.number.toInt())

        val symbolToken = lexicalAnalyzer.nextToken() as SymbolToken
        assertEquals(TokenType.Symbol, symbolToken.tokenType)
        assertEquals(SymbolType.Semicolon, symbolToken.symbolType)
    }

    @Test
    fun testNumber() {
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("12334 323.432"))

        val tokenInteger = lexicalAnalyzer.nextToken() as NumberToken

        assertEquals(TokenType.Integer, tokenInteger.tokenType)
        assertEquals(12334, tokenInteger.number.toInt())

        val tokenDouble = lexicalAnalyzer.nextToken() as NumberToken

        assertEquals(TokenType.Double, tokenDouble.tokenType)
        assertEquals(323.432.toDouble(), tokenDouble.number.toDouble())
    }

    @Test
    fun testKeywordInterval() {
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("Интервал 0.6, 50;"))

        val token:KeywordToken = lexicalAnalyzer.nextToken() as KeywordToken

        assertEquals(TokenType.Keyword, token.tokenType)
        assertEquals(KeywordType.Interval, token.keywordType)

        val tokenDouble = lexicalAnalyzer.nextToken() as NumberToken

        assertEquals(TokenType.Double, tokenDouble.tokenType)
        assertEquals(0.6.toDouble(), tokenDouble.number.toDouble())

        val tokenComma = lexicalAnalyzer.nextToken() as SymbolToken

        assertEquals(TokenType.Symbol, tokenComma.tokenType)
        assertEquals(SymbolType.Comma, tokenComma.symbolType)

        val tokenInteger = lexicalAnalyzer.nextToken() as NumberToken

        assertEquals(TokenType.Integer, tokenInteger.tokenType)
        assertEquals(50, tokenInteger.number.toInt())

        val symbolToken = lexicalAnalyzer.nextToken() as SymbolToken
        assertEquals(TokenType.Symbol, symbolToken.tokenType)
        assertEquals(SymbolType.Semicolon, symbolToken.symbolType)
    }

    @Test
    fun testKeywordStep() {
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("Шаг 0.6;"))

        val token:KeywordToken = lexicalAnalyzer.nextToken() as KeywordToken

        assertEquals(TokenType.Keyword, token.tokenType)
        assertEquals(KeywordType.Step, token.keywordType)

        val tokenDouble = lexicalAnalyzer.nextToken() as NumberToken

        assertEquals(TokenType.Double, tokenDouble.tokenType)
        assertEquals(0.6.toDouble(), tokenDouble.number.toDouble())

        val symbolToken = lexicalAnalyzer.nextToken() as SymbolToken
        assertEquals(TokenType.Symbol, symbolToken.tokenType)
        assertEquals(SymbolType.Semicolon, symbolToken.symbolType)
    }

    @Test
    fun testKeywordValue() {
        val lexicalAnalyzer = DifferLexicalAnalyzer(StringReader("Значения x11 = 10;"))

        val token:KeywordToken = lexicalAnalyzer.nextToken() as KeywordToken

        assertEquals(TokenType.Keyword, token.tokenType)
        assertEquals(KeywordType.Value, token.keywordType)

        val tokenXk = lexicalAnalyzer.nextToken()
        assertEquals(TokenType.Xk, tokenXk.tokenType)
        assertEquals("x11", tokenXk.word.word)

        val equalToken = lexicalAnalyzer.nextToken() as SymbolToken

        assertEquals(TokenType.Symbol, equalToken.tokenType)
        assertEquals(SymbolType.Equal, equalToken.symbolType)

        val numberToken = lexicalAnalyzer.nextToken() as NumberToken

        assertEquals(TokenType.Integer, numberToken.tokenType)
        assertEquals(10, numberToken.number.toInt())

        val symbolToken = lexicalAnalyzer.nextToken() as SymbolToken
        assertEquals(TokenType.Symbol, symbolToken.tokenType)
        assertEquals(SymbolType.Semicolon, symbolToken.symbolType)
    }
}